vllm/worker/model_runner.py

```diff
@@ -165,6 +165,287 @@ def from_broadcasted_tensor_dict(
+            seq_len = min(seq_len, context_len + token_chunk_size)
+
+            # Compute tokens.
+            if is_prompt:
+                tokens = seq_data.get_token_ids()[context_len:seq_len]
+            else:
+                # Optimization. get_token_ids requires the entire copy of
+                # tokens.
+                tokens = [seq_data.get_last_token_id()]
+            if prefix_cache_hit:
+                assert computed_block_nums is not None
+                context_len = len(computed_block_nums) * self.block_size
+                tokens = tokens[context_len:]
+
+            # These are seq_len/context_len capped to the sliding window.
+            # They are passed to decode kernel.
+            # We still need original seq_len/context_len to compute slot
+            # mapping (and input position) below.
+            if is_prompt:
+                curr_sliding_window_block = 0
+                sliding_seq_len = seq_len
+                query_len = seq_len - context_len
+            else:
+                curr_sliding_window_block, sliding_seq_len = (
+                    self._compute_len_for_sliding_window(seq_len))
+                query_len = 1
+
+            self.seq_lens.append(sliding_seq_len)
+            if not is_prompt:
+                self.max_decode_seq_len = max(self.max_decode_seq_len,
+                                              sliding_seq_len)
+            self.query_lens.append(query_len)
+            self.input_tokens.extend(tokens)
+            self.input_positions.extend(list(range(context_len, seq_len)))
+
+            # Intermediate data of the current sequence group for
+            # the attention metadata.
+            token_lens.append(len(tokens))
+            context_lens.append(context_len)
+            decode_seq_lens.append(sliding_seq_len)
+            curr_sliding_window_blocks.append(curr_sliding_window_block)
+            orig_seq_lens.append(seq_len)
+
+        self.attn_metadata_builder.add_seq_group(
+            seq_group_metadata, token_lens, orig_seq_lens, decode_seq_lens,
+            self.query_lens[-n_seqs:], context_lens,
+            curr_sliding_window_blocks, prefix_cache_hit,
+            self.chunked_prefill_enabled)
+
+        # LoRA data.
+        if self.enable_lora:
+            lora_id = seq_group_metadata.lora_int_id
+            for query_len in self.query_lens[-n_seqs:]:
+                if lora_id > 0:
+                    self.lora_requests.add(seq_group_metadata.lora_request)
+                self.lora_index_mapping += [lora_id] * query_len
+                self.lora_prompt_mapping.extend(
+                    [lora_id] *
+                    (query_len if seq_group_metadata.sampling_params
+                     and seq_group_metadata.sampling_params.prompt_logprobs
+                     is not None else 1))
+
+        # Prompt adapter data. Note that when is_prompt=True,
+        # we expect only one sequence in the group.
```

Comment:
assert len(seq_ids) == 1?

Reply:
It has been checked in the beginning of this function.